subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java

```diff
@@ -246,20 +246,64 @@ public void leaveDirectory(DirectorySnapshot directorySnapshot) {
         }
 
         private void visitUnvisitedDirectories() {
+            if (unvisitedDirectories.isEmpty()) {
+                return;
+            }
+            DirectoryVisitState directoryState = new DirectoryVisitState(this);
             DirectorySnapshot unvisited = unvisitedDirectories.poll();
-            String previousPath = this.path;
-            String previousName = this.name;
-            HashCode previousHash = this.hash;
             while (unvisited != null) {
-                this.path = unvisited.getAbsolutePath();
-                this.name = unvisited.getName();
-                this.hash = null;
-                visitor.preDirectory(this);
+                directoryState.path = unvisited.getAbsolutePath();
+                directoryState.name = unvisited.getName();
+                visitor.preDirectory(directoryState);
                 unvisited = unvisitedDirectories.poll();
             }
```

Comment:
If we are improving this code, we could fix the double call to `unvisited = unvisitedDirectories.poll()` too.

Reply:
Done.